-Wdangling-reference warning with newer GCC

### Bug Description

When building MOOSE with gcc 13, instantiations of `MaterialAuxBaseTempl` give me warnings such as:

```
/home/stogrh/git/moose-fresh/framework/build/header_symlinks/MaterialAuxBase.h: In instantiation of 'MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::MaterialAuxBaseTempl(const InputParameters&) [with T = libMesh::DenseMatrix<double>; bool is_ad = false; bool is_functor = false; RT = double]':
/home/stogrh/git/moose-fresh/framework/src/auxkernels/MaterialRealDenseMatrixAux.C:28:42:   required from here
/home/stogrh/git/moose-fresh/framework/build/header_symlinks/MaterialAuxBase.h:47:24: warning: possibly dangling reference to a temporary [-Wdangling-reference]
   47 |   const PropertyType & _prop;
      |                        ^~~~~
/home/stogrh/git/moose-fresh/framework/build/header_symlinks/MaterialAuxBase.h:102:6: note: the temporary was destroyed at the end of the full expression '<lambda closure object>MaterialAuxBaseTempl<libMesh::DenseMatrix<double>, false, false, double>::MaterialAuxBaseTempl(const InputParameters&)::<lambda()>{((MaterialAuxBaseTempl<libMesh::DenseMatrix<double>, false, false, double>*)this)}.MaterialAuxBaseTempl<libMesh::DenseMatrix<double>, false, false, double>::MaterialAuxBaseTempl(const InputParameters&)::<lambda()>()'
   97 |     _prop([this]() -> const auto & {
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~
   98 |       if constexpr (is_functor)
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~
   99 |         return this->template getFunctor<Moose::GenericType<T, is_ad>>("functor");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  100 |       else
      |       ~~~~
  101 |         return this->template getGenericMaterialProperty<T, is_ad>("property");
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  102 |     }()),
      |     ~^~
```

### Steps to Reproduce

Compile MOOSE with `g++ (Ubuntu 13.2.0-23ubuntu4) 13.2.0`, the default compiler on Ubuntu 24.04.1 LTS

### Impact

If there *isn't* a dangling reference bug here, this is just an annoyance, a false positive warning that would at worst encourage developers to ignore more serious warnings in new code.

I don't believe there is a dangling reference bug here - the code looks good, and recent GCC versions have had false positives emitted for that warning.  The false positives I've tested do seem to be fixed for my GCC build, but there are some very similar cases in bug reports.

If there is a dangling reference bug here, of course, that's the worst kind of bug; undefined behavior with "nasal demons" in the best case and silent corruption of data in the worst (which is also the typical) case.

I've got a workaround that I'll put up as a PR for consideration, but with *that* the impact is that we'd be replacing 5 lines of tight C++17-style code with 20 lines of C++98-style template metafunctions.

### [Optional] Diagnostics

_No response_

--- a/framework/include/auxkernels/MaterialAuxBase.h
+++ b/framework/include/auxkernels/MaterialAuxBase.h
@@ -56,7 +56,7 @@ protected:
 
 private:
   /// Helper function to retrieve the property or functor
-  const PropertyType & getPropertyHelper() const;
+  const PropertyType & getPropertyHelper();
 
   /// Multiplier for the material property
   const Real _factor;
@@ -111,7 +111,7 @@ MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::MaterialAuxBaseTempl(
 
 template <typename T, bool is_ad, bool is_functor, typename RT>
 const typename MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::PropertyType &
-MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::getPropertyHelper() const
+MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::getPropertyHelper()
 {
   if constexpr (is_functor)
     return this->template getFunctor<Moose::GenericType<T, is_ad>>("functor");

--- a/framework/include/auxkernels/MaterialAuxBase.h
+++ b/framework/include/auxkernels/MaterialAuxBase.h
@@ -55,6 +55,9 @@ protected:
   Moose::GenericType<T, is_ad> _full_value;
 
 private:
+  /// Helper function to retrieve the property or functor
+  const PropertyType & getPropertyHelper() const;
+
   /// Multiplier for the material property
   const Real _factor;
 
@@ -96,14 +99,7 @@ template <typename T, bool is_ad, bool is_functor, typename RT>
 MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::MaterialAuxBaseTempl(
     const InputParameters & parameters)
   : AuxKernelTempl<RT>(parameters),
-    _prop(
-        [this]() -> const auto &
-        {
-          if constexpr (is_functor)
-            return this->template getFunctor<Moose::GenericType<T, is_ad>>("functor");
-          else
-            return this->template getGenericMaterialProperty<T, is_ad>("property");
-        }()),
+    _prop(getPropertyHelper()),
     _selected_qp(this->isParamValid("selected_qp")
                      ? this->template getParam<unsigned int>("selected_qp")
                      : libMesh::invalid_uint),
@@ -111,6 +107,16 @@ MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::MaterialAuxBaseTempl(
     _offset(this->template getParam<RT>("offset")),
     _current_subdomain_id(this->_assembly.currentSubdomainID())
 {
+}
+
+template <typename T, bool is_ad, bool is_functor, typename RT>
+const typename MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::PropertyType &
+MaterialAuxBaseTempl<T, is_ad, is_functor, RT>::getPropertyHelper() const
+{
+  if constexpr (is_functor)
+    return this->template getFunctor<Moose::GenericType<T, is_ad>>("functor");
+  else
+    return this->template getGenericMaterialProperty<T, is_ad>("property");
 }
 
 template <typename T, bool is_ad, bool is_functor, typename RT>

